lib/sync/validator.go

```diff
@@ -0,0 +1,263 @@
+	"boscoin.io/sebak/lib/common/observer"
+	"boscoin.io/sebak/lib/error"
+	"boscoin.io/sebak/lib/network"
+	"boscoin.io/sebak/lib/node/runner"
+	"boscoin.io/sebak/lib/storage"
+
+	"github.com/inconshreveable/log15"
+)
+
+//TODO(anarcher) another name is Finisher
+
+type BlockValidator struct {
+	network   network.Network
+	storage   *storage.LevelDBBackend
+	commonCfg common.Config
+
+	networkID []byte
+
+	prevBlockWaitTimeout time.Duration // Waiting prev block if is doesn't exist
+	logger               log15.Logger
+}
+
+type BlockValidatorOption func(*BlockValidator)
+
+func NewBlockValidator(nw network.Network, ldb *storage.LevelDBBackend, networkID []byte, cfg common.Config, opts ...BlockValidatorOption) *BlockValidator {
+	v := &BlockValidator{
+		network:              nw,
+		storage:              ldb,
+		networkID:            networkID,
+		prevBlockWaitTimeout: 10 * time.Minute,
+		commonCfg:            cfg,
+
+		logger: NopLogger(),
+	}
+
+	for _, opt := range opts {
+		opt(v)
+	}
+
+	return v
+}
+
+func (v *BlockValidator) Validate(ctx context.Context, syncInfo *SyncInfo) error {
+	exists, err := v.existsBlock(ctx, v.storage, syncInfo.BlockHeight)
+	if err != nil {
+		return err
+	}
+	if exists == true {
+		v.logger.Info("This block exists", "height", syncInfo.BlockHeight)
+		return nil
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		if err := v.validate(ctx, syncInfo); err != nil {
+			return err
+		}
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		return v.finishBlock(ctx, syncInfo)
+	}
+}
+
+func (v *BlockValidator) validate(ctx context.Context, syncInfo *SyncInfo) error {
+	//Waiting to get prev block for runner.ValidateTx
+	prevBlk, err := v.getPrevBlock(ctx, syncInfo.BlockHeight)
+	if err != nil {
+		return err
+	}
+
+	if err := v.validateTxs(ctx, syncInfo); err != nil {
+		return err
+	}
+
+	if err := v.validateBlock(ctx, syncInfo, prevBlk); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (v *BlockValidator) finishBlock(ctx context.Context, syncInfo *SyncInfo) error {
+	ts, err := v.storage.OpenTransaction()
+	if err != nil {
+		return err
+	}
+
+	if exists, err := v.existsBlock(ctx, ts, syncInfo.BlockHeight); err != nil {
+		ts.Discard()
+		return err
+	} else if exists == true {
+		v.logger.Info("This block exists", "height", syncInfo.BlockHeight)
+		return nil
+	}
+
+	//TODO(anarcher): using leveldb.Tx or leveldb.Batch?
+	blk := *syncInfo.Block
+	if err := blk.Save(ts); err != nil {
+		if err == errors.ErrorBlockAlreadyExists {
+			return nil
+		}
+		return err
+	}
+
+	if err := runner.FinishTransactions(blk, syncInfo.Txs, ts); err != nil {
+		ts.Discard()
+		return err
+	}
+
+	ptx := syncInfo.Ptx
+	if err := runner.FinishProposerTransaction(ts, blk, *ptx, v.logger); err != nil {
+		ts.Discard()
+		return err
+	}
+
+	v.logger.Debug(fmt.Sprintf("finish to sync block height: %v", syncInfo.BlockHeight), "height", syncInfo.BlockHeight, "hash", blk.Hash)
```

Comment:
Redundant `Sprintf`

Reply:
No, I want to log like `finish to sync block height: 123        height=123`.  It is easy to find the height in msg.